pkg/ingester/circuitbreaker.go

```diff
@@ -0,0 +1,231 @@
+	f.DurationVar(&cfg.InitialDelay, prefix+"initial-delay", 0, "How long the circuit breaker should wait to start up after the corresponding ingester started. During that time both failures and successes will not be counted.")
+	f.DurationVar(&cfg.PushTimeout, prefix+"push-timeout", defaultPushTimeout, "How long is execution of ingester's Push supposed to last before it is reported as timeout in a circuit breaker. This configuration is used for circuit breakers only, and timeout expirations are not reported as errors")
+}
+
+type circuitBreaker struct {
+	cfg     CircuitBreakerConfig
+	logger  log.Logger
+	metrics *circuitBreakerMetrics
+	active  atomic.Bool
+	cb      circuitbreaker.CircuitBreaker[any]
+
+	// testRequestDelay is needed for testing purposes to simulate long lasting requests
+	testRequestDelay time.Duration
+}
+
+func newCircuitBreaker(cfg CircuitBreakerConfig, isActive bool, logger log.Logger, registerer prometheus.Registerer) *circuitBreaker {
+	active := atomic.NewBool(isActive)
+	cb := circuitBreaker{
+		cfg:    cfg,
+		logger: logger,
+		active: *active,
+	}
+
+	circuitBreakerTransitionsCounterFn := func(metrics *circuitBreakerMetrics, state circuitbreaker.State) prometheus.Counter {
+		return metrics.circuitBreakerTransitions.WithLabelValues(state.String())
+	}
+
+	cbBuilder := circuitbreaker.Builder[any]().
+		WithFailureThreshold(cfg.FailureThresholdPercentage).
+		WithDelay(cfg.CooldownPeriod).
+		OnClose(func(event circuitbreaker.StateChangedEvent) {
+			circuitBreakerTransitionsCounterFn(cb.metrics, circuitbreaker.ClosedState).Inc()
+			level.Info(logger).Log("msg", "circuit breaker is closed", "previous", event.OldState, "current", event.NewState)
+		}).
+		OnOpen(func(event circuitbreaker.StateChangedEvent) {
+			circuitBreakerTransitionsCounterFn(cb.metrics, circuitbreaker.OpenState).Inc()
+			level.Warn(logger).Log("msg", "circuit breaker is open", "previous", event.OldState, "current", event.NewState)
+		}).
+		OnHalfOpen(func(event circuitbreaker.StateChangedEvent) {
+			circuitBreakerTransitionsCounterFn(cb.metrics, circuitbreaker.HalfOpenState).Inc()
+			level.Info(logger).Log("msg", "circuit breaker is half-open", "previous", event.OldState, "current", event.NewState)
+		})
+
+	if cfg.testModeEnabled {
+		// In case of testing purposes, we initialize the circuit breaker with count based failure thresholding,
+		// since it is more deterministic, and therefore it is easier to predict the outcome.
+		cbBuilder = cbBuilder.WithFailureThreshold(cfg.FailureThresholdPercentage)
+	} else {
+		// In case of production code, we prefer time based failure thresholding.
+		cbBuilder = cbBuilder.WithFailureRateThreshold(cfg.FailureThresholdPercentage, cfg.FailureExecutionThreshold, cfg.ThresholdingPeriod)
+	}
+
+	cb.cb = cbBuilder.Build()
+	cb.metrics = newCircuitBreakerMetrics(registerer, cb.cb.State)
+	return &cb
+}
+
+func isFailure(err error) bool {
```

Comment:
I am slightly concerned about polluting the entire `pkg/ingester` namespace with all these circuit-breaker specific naminig: `ingester.isFailure(err)` for example, or `defaultPushTimeout` or `resultOpen`: all of them are circuit-breaker specific but their names don't represent that.

Have you considered prefixing them, or maybe even moving all this code to a sub-package?

Reply:
Thank you.